sympy/core/basic.py

```diff
@@ -1480,55 +1487,40 @@ def replace(self, query, value, map=False, simultaneous=True, exact=None):
+            if args is not None:
+                if args:
+                    newargs = tuple([walk(a, F) for a in args])
+                    if args != newargs:
+                        rv = rv.func(*newargs)
+                        if simultaneous:
+                            # if rv is something that was already
+                            # matched that was changed then skip
+                            # applying F again
+                            for i, e in enumerate(args):
+                                if rv == e and e != newargs[i]:
+                                    return rv
+                rv = F(rv)
+            return rv
+
+
+        mapping = []  # changes that took place
 
         def rec_replace(expr):
             result = _query(expr)
             if result or result == {}:
-                new = _value(expr, result)
-                if new is not None and new != expr:
-                    mapping[expr] = new
-                    if simultaneous:
-                        # don't let this change during rebuilding;
-                        # XXX this may fail if the object being replaced
-                        # cannot be represented as a Dummy in the expression
-                        # tree, e.g. an ExprConditionPair in Piecewise
-                        # cannot be represented with a Dummy
-                        com = getattr(new, 'is_commutative', True)
-                        if com is None:
-                            com = True
-                        d = Dummy('rec_replace', commutative=com)
-                        mask.append((d, new))
-                        expr = d
-                    else:
-                        expr = new
+                v = _value(expr, result)
+                if v is not None and v != expr:
+                    if map:
+                        mapping.append((expr, v))
+                    expr = v
             return expr
 
-        rv = bottom_up(self, rec_replace, atoms=True)
-
-        # restore original expressions for Dummy symbols
-        if simultaneous:
-            mask = list(reversed(mask))
-            for o, n in mask:
-                r = {o: n}
-                # if a sub-expression could not be replaced with
-                # a Dummy then this will fail; either filter
-                # against such sub-expressions or figure out a
-                # way to carry out simultaneous replacement
-                # in this situation.
-                rv = rv.xreplace(r)  # if this fails, see above
-
-        if not map:
-            return rv
-        else:
-            if simultaneous:
-                # restore subexpressions in mapping
-                for o, n in mask:
-                    r = {o: n}
-                    mapping = {k.xreplace(r): v.xreplace(r)
-                        for k, v in mapping.items()}
-            return rv, mapping
+        rv = walk(self, rec_replace)
+        return (rv, dict(mapping)) if map else rv
```

Comment:
```suggestion
        return (rv, mapping) if map else rv
```

Reply:
At one point I was using the ordered replacements to create RV. You are right, this can go back to dict.